azurecontainerapps/azurecontainerapps.ts

```diff
@@ -30,200 +23,507 @@ export class azurecontainerapps {
             tl.mkdirP(cwd);
             tl.cd(cwd);
 
-            // Set build variables used later for default values
-            const buildId = tl.getVariable('Build.BuildId');
-            const buildNumber = tl.getVariable('Build.BuildNumber');
+            // Validate that the arguments provided can be used for one of the supported scenarios
+            this.validateSupportedScenarioArguments();
 
-            // Set up array to store optional arguments for the 'az containerapp up' command
-            const optionalCmdArgs: string[] = [];
+            // Set up the Azure CLI to be used for this task
+            this.setupAzureCli();
 
-            // Get the path to the application source to build and run, if provided
-            const appSourcePath: string = tl.getInput('appSourcePath', false);
+            // Set up the resources required to deploy a Container App
+            this.setupResources();
 
-            // Get the name of the ACR instance to push images to, if provided
-            const acrName: string = tl.getInput('acrName', false);
+            // If an Azure Container Registry name was provided, try to authenticate against it
+            if (!util.isNullOrEmpty(this.acrName)) {
+                await this.authenticateAzureContainerRegistryAsync();
+            }
 
-            // Get the previously built image to deploy, if provided
-            let imageToDeploy: string = tl.getInput('imageToDeploy', false);
+            // If the application source was provided, build a runnable application image from it
+            if (!util.isNullOrEmpty(this.appSourcePath)) {
+                await this.buildAndPushImageAsync();
+            }
 
-            // Ensure that acrName is also provided if appSourcePath is provided
-            if (!util.isNullOrEmpty(appSourcePath) && util.isNullOrEmpty(acrName)) {
-                tl.error(tl.loc('MissingAcrNameMessage'));
-                throw Error(tl.loc('MissingAcrNameMessage'));
+            // If no application source was provided, set up the scenario for deploying an existing image
+            if (util.isNullOrEmpty(this.appSourcePath)) {
+                this.setupExistingImageScenario();
```

Comment:
So if there is no app source path, our validation checks above verified that we at least are passing in an imageToDeploy or a yaml config file, either of which will contain an existing image, and therefore we can be sure that we are using a pre-existing image instead of building our own?

Reply:
Exactly -- if no value for `appSourcePath` is provided, that means the user doesn't want the task to produce an image from their application code, and that they want to use an image that was previously built (either earlier in their pipeline, or previously published to a registry). This value is then provided as the `imageToDeploy` argument, or it's specified in the YAML configuration file's `properties.template.containers.image` field. If neither the `imageToDeploy` nor the `yamlConfigPath` arguments are provided, then the validation will have previously failed.